experimental/trusted_shuffler/client/src/main.rs

```diff
@@ -32,27 +35,106 @@ pub struct Opt {
-    info!(
-        "Received response: {} in {:?}",
-        parsed_response, request_duration
-    );
+    let expected_qps = opt.qps;
+    let rounds = opt.rounds;
+
+    // Start with giving one second iterations.
+    let start_time = Instant::now();
+
+    // Header of table printed by logger.
+    log::info!("phase,id,request_elapsed_in_ms,response_elapsed_in_ms");
+
+    let mut clients = vec![];
+
+    for (n, p) in Sieve::new()
+        .iter()
+        .take((expected_qps * rounds) as usize)
+        .enumerate()
+    {
+        let server_url = opt.server_url.clone();
+
+        // We currently spawn a new client for every new request. Alternatively we could re-use the
+        // client for every round.
+        clients.push(tokio::spawn(async move {
+            let request = format!("{}", p);
+            let request_start = start_time.elapsed();
+
+            let url = format!("{}/request", server_url);
+            let response = send_request(&url, Method::POST, request.as_bytes())
+                .await
+                .context("Couldn't receive response");
+
+            let request_duration = start_time.elapsed();
+
+            if let Ok(response) = response {
+                // We don't check whether response is exepected response.
+                let _parsed_response =
+                    String::from_utf8(response).context("Couldn't decode response body");
+
+                log::info!(
+                    "client,{},{},{}",
+                    p,
+                    request_start.as_millis(),
+                    request_duration.as_millis(),
+                );
+            }
+        }));
+
+        // Compute current round from queries already sent.
+        let current_round = n as u32 / expected_qps;
+
+        // Adapt the sleep_time depending on how many queries still have to fit.
+        let queries_sent_this_round = (n + 1) as u32 % expected_qps;
+        let queries_still_to_fit_this_round = expected_qps - queries_sent_this_round;
+
+        let time_elapsed_this_round = Duration::checked_sub(
+            start_time.elapsed(),
+            Duration::from_secs(current_round as u64),
+        )
+        .expect("The time elapsed is shorter than current round.");
+
+        // If there is no time left, we set time_still_left to 0.
+        let time_still_left_this_round =
+            Duration::checked_sub(Duration::from_secs(1), time_elapsed_this_round)
+                .unwrap_or(Duration::from_secs(0));
+
+        // If there are no queries still to fit in this round and there is time left in the round,
+        // we sleep until the second has elapsed.
+        let delta = time_still_left_this_round
+            .checked_div(queries_still_to_fit_this_round)
+            .unwrap_or(
+                Duration::checked_sub(Duration::from_secs(1), time_elapsed_this_round)
+                    .unwrap_or(Duration::from_secs(0)),
+            );
+
+        sleep(delta).await;
```

Comment:
Would it work to spawn a task every exactly 1s that essentially performs the body of this loop, instead of trying to calculate deltas? Or alternatively start the 1s sleep in the background at the beginning of this loop, and then await for it here?